src/accelerate/accelerator.py

```diff
@@ -1749,6 +1748,40 @@ def _prepare_ipex(self, *args):
                 result[i] = optimizer
         return tuple(result)
 
+    def _prepare_msamp(self, *args):
+        if not is_msamp_available():
+            raise ImportError(
+                "MS-AMP was not found on your system. Please ensure that MS-AMP is available "
+                " or choose `'te'` as the backend for FP8 mixed precision training."
+            )
+        else:
+            import msamp
+
+        model, optimizer = None, None
+        num_models, num_optimizers = 0, 0
+        result = [obj for obj in args]
+        for obj in result:
+            if isinstance(obj, torch.nn.Module):
+                model = obj
+                num_models += 1
+            elif isinstance(obj, (torch.optim.Optimizer)):
+                optimizer = obj
+                num_optimizers += 1
+        if optimizer is None and model is None:
+            raise ValueError("You must pass a model and an optimizer to `accelerate.prepare()` when using MS-AMP.")
+        elif num_models > 1 or num_optimizers > 1:
+            raise ValueError(
+                f"You can't use multiple models ({num_models}) or optimizers {num_optimizers} with MS-AMP."
+            )
+        else:
```

Comment:
There is also the case of `num_models==0` and `num_optimizers==1`, which is invalid. I guess MS-AMP will error in that case, but could be worth to check explicitly and give a nice error message to the user (probably they try preparing the components separately if that happens).

Reply:
Resolved, we now raise the err if either `optimizer` or `model` is `None`, since while msamp can magically create `Adam` we do not want to advertise Accelerate as doing so